Add support for linear system convergence

GiudGiud opened the pull request:
builds on top of https://github.com/idaholab/moose/pull/30592

also fixes problems with physics exodus restart, and iteration adaptive DT with linear systems

GiudGiud committed:
Add support for iteration adaptive dt with linear FV
refs #30681

GiudGiud committed:
Address reviews:
- implement iteratieAdaptiveDT for linear-only solves
refs #30681

GiudGiud committed:
Add a test for iterative time stepping using linear systems
refs #30681

GiudGiud committed:
Add a reference attribute for compatiblity with derived classes

GiudGiud committed:
Enable exodus restart from Physics for linear systems

GiudGiud committed:
Use convergence objects in linear systems-only solves
refs #21600

GiudGiud committed:
Add a complex convergence behavior to test time step growing and shrinking with linear systems
refs #30681

GiudGiud commented:
Test failures unrelated.
See comments in https://github.com/idaholab/moose/pull/30592 for anyone reviewing this

lindsayad commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -8991,7 +8994,36 @@ FEProblemBase::getNonlinearConvergenceNames() const
+bool
+FEProblemBase::hasLinearConvergenceObjects() const
+{
+  if (_set_linear_convergence_names)
```
Do you need this boolean member? Can't you just return `_linear_convergence_names.size()`?

GiudGiud replied:
the problem is whether we want to detect whether it has been set and is empty, or whether it has not been set and that s why it's empty.

This is what the nonlinear convergence are doing

GiudGiud replied:
The empty case for nonlinear is usually temporary, before we add the default. 

I could see the linear ones getting a default at some point too

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -354,6 +354,55 @@ petscNonlinearConverged(SNES /*snes*/,
+  // TODO: add a dedicated flag once support is fully implemented.
+  // Right now, setting objects to execute on this flag would be ignored except in the
+  // linear-system-only use case.
+  problem.execute(EXEC_NONLINEAR_CONVERGENCE);
```
add additional exec flag instead of using the `EXEC_LINEAR_CONVERGENCE`

GiudGiud replied:
I assume you mean EXEC_NONLINEAR_CONVERGENCE.
I added EXEC_LINEAR_CONVERGENCE

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -354,6 +354,55 @@ petscNonlinearConverged(SNES /*snes*/,
+  Convergence::MooseConvergenceStatus status;
+  // TODO: This API could become about the "solverConvergenceCheck" to apply to both linear and
+  // nonlinear system convergence checks.
+  if (problem.getFailNextNonlinearConvergenceCheck())
```
Either rename it `getFailNextConvergenceCheck` (or something) or add new API. I don't want to use conceptually wrong names throughout this function. Same comment applies to all uses of 'nonlinear' in this function

lindsayad replied:
The TODO comment should be gone before this is merged

GiudGiud replied:
gone. I think the APIs can be determined when we enable linear-within-nonlinear convergence checks. It is using dedicated APIs now, with the nonlienar APIs forwarding to the new more general APIs

lindsayad reviewed (changes requested):
We can drop all the NS stuff right?

GiudGiud committed:
Address Alex review:
- add dedicated API and execute on flag for the linear system convergence check
- use a common API for the 'fail' convergence check
refs #21600

GiudGiud commented:
NS stuff dropped by rebasing

joshuahansel commented on framework/src/utils/PetscSupport.C:
```diff
@@ -511,7 +557,16 @@ petscSetDefaults(FEProblemBase & problem)
+    }
+
+    // We dont set the KSP defaults here because they seem to clash with the linear solve parameters
+    // set in FEProblemBase::solveLinearSystem
```
I guess this must work, but I don't understand it. So there is no default linear convergence object being added (`problem.hasLinearConvergenceObjects() == false)`, so in most cases today, it just looks like `petscSetKSPDefaults(problem, ksp)` has no effect (we could delete this line without any of your other changes)?

By the way, is this only for the `solve_type = LINEAR` solves, or also for the linear solves inside nonlinear solves?

lindsayad commented on framework/src/utils/PetscSupport.C:
```diff
@@ -511,7 +557,16 @@ petscSetDefaults(FEProblemBase & problem)
-    petscSetKSPDefaults(problem, ksp);
+
+    if (problem.hasLinearConvergenceObjects())
+    {
+      LibmeshPetscCallA(
+          lin_sys.comm().get(),
+          KSPSetConvergenceTest(ksp, petscLinearConverged, &problem, LIBMESH_PETSC_NULLPTR));
+    }
```
```suggestion
      LibmeshPetscCallA(
          lin_sys.comm().get(),
          KSPSetConvergenceTest(ksp, petscLinearConverged, &problem, LIBMESH_PETSC_NULLPTR));
```

GiudGiud commented:
Test failures (threaded eigen problem notably) unrelated
